File: maximumSwap.py
#===solution on the second greedy way https://leetcode.com/articles/maximum-swap/
class Solution(object):
    def maximumSwap(self, num):
        """
        :type num: int
        :rtype: int
        """
        A = map(int, str(num)) #3455 => [3,4,5,5]
        last = {x: i for i, x in enumerate(A)} #{3: 0, 4:1, 5:3}
        for i, n in enumerate(A):
            for nextBig in xrange(9, n, -1):
                if last.get(nextBig, None) > i:
                    A[i], A[last[nextBig]] = A[last[nextBig]], A[i]
                    return int("".join(map(str, A)))
        return num

File: maximumSwap.kt
// editorial solution 4 https://leetcode.com/problems/maximum-swap/editorial/?envType=company&envId=facebook&favoriteSlug=facebook-thirty-days
//basically we scan from right to left, save the highest digit index so far, and whenever we saw a digit to the left to be lower than our highest, we update the swap candidate pair
// if no pair is found, we know the swap is unnecessary
class Solution {
    fun maximumSwap(num: Int): Int {
        val numstr = num.toString().toCharArray()
        val length = numstr.size
        var maxNumberIdxFromRight = -1
        var swapL = -1
        var swapR = -1
        for(i in length-1 downTo 0){
            val currInt = numstr[i].digitToInt()
            if(maxNumberIdxFromRight == -1 || currInt > numstr[maxNumberIdxFromRight].digitToInt())
                maxNumberIdxFromRight = i
            else if (currInt < numstr[maxNumberIdxFromRight].digitToInt()) {
                swapL = i
                swapR = maxNumberIdxFromRight
            }
        }

        if(swapL != -1 && swapR != -1){
            val temp = numstr[swapR]
            numstr[swapR] = numstr[swapL]
            numstr[swapL] = temp
        }
        return numstr.joinToString("").toInt()
    }
}
